yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -213,8 +216,11 @@ private[yarn] class YarnAllocator(
    * Deal with any containers YARN has granted to us by possibly launching executors in them.
    *
    * This must be synchronized because variables read in this method are mutated by other methods.
+   *
+   * Returns a list of executor loss reasons discovered by the allocator, which can then be forwarded
+   * to the driver by the calling ApplicationMaster.
    */
-  def allocateResources(): Unit = synchronized {
+  def allocateResources(): Option[Map[String, ExecutorLossReason]] = synchronized {
```

Comment:
Why not return an empty map instead of an option. It would reduce some complexity when reading it in launchReporterThread()